### PR TITLE
PreValidate virtual method is added

### PR DESCRIPTION
**Pull request by AdemCatamak**
Issue #801 

I added the virtual PreValidate method. This method will work before the PreValidate method and will decide whether the Validate / ValidateAync method will work according to the return value.

In this method, operations on the error list of the ValidationResult object can be performed

I check `Allow edits from maintainers`. Please do not hesitate to make changes before approve of pr

### CODE REVIEW

File: src/FluentValidation/AbstractValidator.cs [hunk 1]
```diff
@@ -374,5 +397,16 @@ public abstract class AbstractValidator<T> : IValidator<T>, IEnumerable<IValidat
...
 		}
+
+		/// <summary>
+		/// If return true, Validate/ValidateAsync will be executed.
+		/// If return false, result is returned immediatly as result of Validate/ValidateAsync.
+		/// </summary>
+		/// <param name="context"></param>
+		/// <param name="result"></param>
+		/// <returns></returns>
+		protected virtual bool  PreValide(ValidationContext<T> context, ValidationResult result) {
```
Comment (JeremySkinner) [inline src/FluentValidation/AbstractValidator.cs]: Typo. Should be `PreValidate`

File: src/FluentValidation/AbstractValidator.cs [hunk 2]
```diff
@@ -113,9 +113,20 @@ public abstract class AbstractValidator<T> : IValidator<T>, IEnumerable<IValidat
...
+			var result = new ValidationResult();
+			bool shouldContinue = PreValide(context, result);
+			if (!shouldContinue) {
+				return result;
+			}
+
 			EnsureInstanceNotNull(context.InstanceToValidate);
 			var failures = NestedValidators.SelectMany(x => x.Validate(context));
-			return new ValidationResult(failures);
+			foreach (ValidationFailure validationFailure in failures.Where(failure => failure != null)) {
```
Comment (JeremySkinner) [inline src/FluentValidation/AbstractValidator.cs]: Use `var` please.

File: src/FluentValidation/AbstractValidator.cs [hunk 3]
```diff
@@ -136,7 +154,12 @@ public abstract class AbstractValidator<T> : IValidator<T>, IEnumerable<IValidat
 				NestedValidators
 				.Select(v => v.ValidateAsync(context, cancellation)
 				.Then(fs => failures.AddRange(fs), runSynchronously: true, cancellationToken: cancellation)), cancellation)
-				.Then(() => new ValidationResult(failures), cancellation);
+				.Then(() => {
+					      foreach (ValidationFailure validationFailure in failures.Where(failure => failure != null)) {
```
Comment (JeremySkinner) [inline src/FluentValidation/AbstractValidator.cs]: `var` here too please.

**Comment by AdemCatamak** [comment]
I spent your time for typo and style issues. I am sorry about that

I think this time I fix all of them :)

**Comment by JeremySkinner** [comment]
I've merged this in and made 2 minor changes:
- Updated the doc comment 
- Moved the `EnsureInstanceNotNull` check to happen before `PreValidate`

**Comment by AdemCatamak** [comment]
Thanks for effort 👍

**Pull request merged**